### PR TITLE
feat: language switch

### PR DESCRIPTION
**Pull request by hyoban**
### <samp>🤖 Generated by Copilot at 1efa52b</samp>

Added a `LanguageSwitch` component to the home page to allow users to choose their preferred language. Updated the `Menu` component to use a better type for placement. Added support for reading and setting the `preferred_language` cookie in the `getAcceptLang` function and the `LanguageSwitch` component. Created a `getAvailableLanguages` function to get the list of supported languages from `i18next`.

### <samp>🤖 Generated by Copilot at 1efa52b</samp>

> _Sing, O Muse, of the skillful code review_
> _That brought new features to the web app's view_
> _How the `LanguageSwitch` component shone_
> _Like a bright star among the common ones_

### WHY

I'm used to setting my computer to English, then I can't adjust the language preference to Chinese like on other sites

### <samp>🤖 Generated by Copilot at 1efa52b</samp>

* Create and render `LanguageSwitch` component to allow user to change preferred language ([link](https://github.com/Crossbell-Box/xLog/pull/903/files?diff=unified&w=0#diff-d66da1b2249038dc5d7672bcb640b2c0d9317e6ab32b36c0e0b1f6ccb9386123R3), [link](https://github.com/Crossbell-Box/xLog/pull/903/files?diff=unified&w=0#diff-d66da1b2249038dc5d7672bcb640b2c0d9317e6ab32b36c0e0b1f6ccb9386123R48), [link](https://github.com/Crossbell-Box/xLog/pull/903/files?diff=unified&w=0#diff-20a28c95d79c95f979b2a8cbe057b075a4cd414108fa755f20bfb59ceeb188f8R1-R38))
* Use `cookies` function and `preferred_language` cookie to store and access user's preferred language in `getAcceptLang` function ([link](https://github.com/Crossbell-Box/xLog/pull/903/files?diff=unified&w=0#diff-a92d3789380a0c764653a8dd9ea8f9233fedc5459dfa8c64fb63cfe79cdbc4edL1-R1), [link](https://github.com/Crossbell-Box/xLog/pull/903/files?diff=unified&w=0#diff-a92d3789380a0c764653a8dd9ea8f9233fedc5459dfa8c64fb63cfe79cdbc4edR8-R9), [link](https://github.com/Crossbell-Box/xLog/pull/903/files?diff=unified&w=0#diff-a92d3789380a0c764653a8dd9ea8f9233fedc5459dfa8c64fb63cfe79cdbc4edL14-R22))
* Use `Placement` type from `@floating-ui/react` for `placement` prop of `Menu` component ([link](https://github.com/Crossbell-Box/xLog/pull/903/files?diff=unified&w=0#diff-83f41a1bc395f5576bafba57dfb869b03577e0e558235312013169c29bb6a1bbL4-R4), [link](https://github.com/Crossbell-Box/xLog/pull/903/files?diff=unified&w=0#diff-83f41a1bc395f5576bafba57dfb869b03577e0e558235312013169c29bb6a1bbL22-R22))
* Create `getAvailableLanguages` function to return supported languages from `i18next` library ([link](https://github.com/Crossbell-Box/xLog/pull/903/files?diff=unified&w=0#diff-c3551dac02c8d487d5d5650caea0374963d1096683de89e387a29c6290695961R36-R42))

### CLAIM REWARDS

For first-time contributors, please leave your xLog address and Discord ID below to claim your rewards.

### CODE REVIEW

**Comment by hyoban** [comment]
ScreenShot 2023-08-13 18 26 52

File: src/components/common/LanguageSwitch.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import { getAvailableLanguages } from "~/lib/i18n/client"
+
+import { Menu } from "../ui/Menu"
+
+export function LanguageSwitch() {
+  const languages = getAvailableLanguages()
+
+  const changeLanguage = (language: string) => {
+    document.cookie = `preferred_language=${language};${document.cookie}`
```
Comment (DIYgod) [inline src/components/common/LanguageSwitch.tsx:11]: Every time the language is switched, `preferred_language` field will be repeatedly added.

File: src/lib/i18n/client.ts [hunk 2]
```diff
@@ -33,3 +33,10 @@ export function useTranslation(ns: string = defaultNS) {
 }
 
 export const Trans = TransW
+
+export function getAvailableLanguages() {
+  const supportedLngs = i18next.options.supportedLngs
+  return typeof supportedLngs === "object"
+    ? supportedLngs.filter((i) => i !== "cimode")
```
Comment (DIYgod) [inline src/lib/i18n/client.ts:40]: What is `cimode` used for?

**Review by DIYgod — CHANGES_REQUESTED** [review]

File: src/components/common/LanguageSwitch.tsx (hunk 1 above)
Comment (hyoban) [reply src/components/common/LanguageSwitch.tsx:11]: From what I've seen, it doesn't seem to be, it looks like a direct replacement. The following is the log when i change the language.

ScreenShot 2023-08-16 10 12 43

File: src/lib/i18n/client.ts (hunk 2 above)
Comment (hyoban) [reply src/lib/i18n/client.ts:40]: > If set to 'cimode', the output text will be the key.

https://www.i18next.com/overview/configuration-options#languages-namespaces-resources

**Comment by DIYgod** [comment]
I have some suggestions for user experience:

1. Identify which option is currently being used.
2. Use more human-readable descriptions for option text.
3. Provide an "auto" option to delete cookies and use automatic language detection.
4. Also provide the same switch button at the bottom of the [site footer](https://github.com/Crossbell-Box/xLog/blob/dev/src/components/site/SiteFooter.tsx).

**Comment by hyoban** [comment]
I have modified the code according to your opinion, please review it when you have time

**Comment by DIYgod** [comment]
Thanks so much!

**Pull request merged**